salt-api: Prevent impersonation of ServiceAccounts using Bearer tokens

gdemonet opened the pull request:
**Note for reviewers**:

This PR adds some changes to Basic auth handling, and does so in different manners:

- in 2.4, Basic auth support in SaltAPI is maintained (through the `password` value for `POST /login`), but now uses plaintext password instead of the `user:password` base64-encoded token
- in 2.5 and upwards, Basic auth support in SaltAPI is entirely removed, and logic for having multiple handlers based on authentication method is removed - as such, we also remove this particular logic from our clients, namely `storage-operator` and `metalk8s-ui`


<details>
  <summary>Commits</summary>

### salt-api: Rework Basic auth handling

Instead of relying on the strange assumption that the /login 'password'
argument is actually the full 'user:password' base64-encoded token, we
adjust the storage-operator and metalk8s-ui to pass in this password in
cleartext.

This commit also updates the `_auth_basic` handler to rely on the same
SelfSubjectAccessReview API as the  `_groups_basic` handler did.

---

### salt-api: Prevent SA impersonation

When using ServiceAccount, we were not checking if the provided Bearer
`token` actually belonged to the claimed `username`. This is now done,
and we add tests to prevent regressions.

NOTE: storage-operator username, from SaltAPI PoV, is now
'system:serviceaccount:kube-system:storage-operator'.

Fixes: #2634

---

### tests: Add positive check for auth success

We only assessed if an authentication attempt failed, while tried to
check the actual permissions for an authn assumed successful.
To make errors more explicit, we add a `then` check that the
"authentication succeeds".

See: #2634

</details>

gdemonet committed:
salt-api: Rework Basic auth handling

Instead of relying on the strange assumption that the /login 'password'
argument is actually the full 'user:password' base64-encoded token, we
adjust the storage-operator and metalk8s-ui to pass in this password in
cleartext.

This commit also updates the `_auth_basic` handler to rely on the same
SelfSubjectAccessReview API as the  `_groups_basic` handler did.

bert-e commented:
# Hello gdemonet,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Branches have diverged



This pull request's source branch `bugfix/2634-impersonation-salt-api` has diverged from
`development/2.6` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.6` into `bugfix/2634-impersonation-salt-api`
* Rebase `bugfix/2634-impersonation-salt-api` onto `origin/development/2.6`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.5/bugfix/2634-impersonation-salt-api` with contents from `bugfix/2634-impersonation-salt-api`
and `development/2.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.5/bugfix/2634-impersonation-salt-api origin/development/2.5
 $ git merge origin/bugfix/2634-impersonation-salt-api
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.5/bugfix/2634-impersonation-salt-api
```

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/2634-impersonation-salt-api` into
`development/2.4`
* [w/2.5/bugfix/2634-impersonation-salt-api](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/bugfix/2634-impersonation-salt-api) will be merged into `development/2.5` 
* [w/2.6/bugfix/2634-impersonation-salt-api](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/bugfix/2634-impersonation-salt-api) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality

gdemonet commented:
_Need to fix the storage-operator unit tests_

gdemonet committed:
tests: Add positive check for auth success

We only assessed if an authentication attempt failed, while tried to
check the actual permissions for an authn assumed successful.
To make errors more explicit, we add a `then` check that the
"authentication succeeds".

See: #2634

bert-e commented:
# History mismatch



Merge commit #f05b7617a88dbbd4466923c86857d30a7e23cbef on the integration branch
`w/2.6/bugfix/2634-impersonation-salt-api` is merging a branch which is neither the current
branch `bugfix/2634-impersonation-salt-api` nor the development branch
`development/2.6`.

It is likely due to a rebase of the branch `bugfix/2634-impersonation-salt-api` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

bert-e commented:
# History mismatch



Merge commit #c5a5119e7136d72210222334877a6d3407ca6815 on the integration branch
`w/2.5/bugfix/2634-impersonation-salt-api` is merging a branch which is neither the current
branch `bugfix/2634-impersonation-salt-api` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `bugfix/2634-impersonation-salt-api` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.5/bugfix/2634-impersonation-salt-api` with contents from `bugfix/2634-impersonation-salt-api`
and `development/2.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.5/bugfix/2634-impersonation-salt-api origin/development/2.5
 $ git merge origin/bugfix/2634-impersonation-salt-api
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.5/bugfix/2634-impersonation-salt-api
```

bert-e commented:
# History mismatch



Merge commit #193f4fb5a98c9a1b7d03a956a5e6014996c30a6f on the integration branch
`w/2.5/bugfix/2634-impersonation-salt-api` is merging a branch which is neither the current
branch `bugfix/2634-impersonation-salt-api` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `bugfix/2634-impersonation-salt-api` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.5/bugfix/2634-impersonation-salt-api` with contents from `bugfix/2634-impersonation-salt-api`
and `development/2.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.5/bugfix/2634-impersonation-salt-api origin/development/2.5
 $ git merge origin/bugfix/2634-impersonation-salt-api
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.5/bugfix/2634-impersonation-salt-api
```

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/2634-impersonation-salt-api` into
`development/2.4`
* [w/2.5/bugfix/2634-impersonation-salt-api](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/bugfix/2634-impersonation-salt-api) will be merged into `development/2.5` 
* [w/2.6/bugfix/2634-impersonation-salt-api](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/bugfix/2634-impersonation-salt-api) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality

gdemonet committed:
salt-api: Prevent SA impersonation

When using ServiceAccount, we were not checking if the provided Bearer
`token` actually belonged to the claimed `username`. This is now done,
and we add tests to prevent regressions.

NOTE: storage-operator username, from SaltAPI PoV, is now
'system:serviceaccount:kube-system:storage-operator'.

Fixes: #2634

bert-e commented:
# History mismatch



Merge commit #e59f57e7ba0927d41cfb16e7382fe159ad75fe99 on the integration branch
`w/2.5/bugfix/2634-impersonation-salt-api` is merging a branch which is neither the current
branch `bugfix/2634-impersonation-salt-api` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `bugfix/2634-impersonation-salt-api` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.5/bugfix/2634-impersonation-salt-api` with contents from `bugfix/2634-impersonation-salt-api`
and `development/2.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.5/bugfix/2634-impersonation-salt-api origin/development/2.5
 $ git merge origin/bugfix/2634-impersonation-salt-api
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.5/bugfix/2634-impersonation-salt-api
```

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/2634-impersonation-salt-api` into
`development/2.4`
* [w/2.5/bugfix/2634-impersonation-salt-api](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/bugfix/2634-impersonation-salt-api) will be merged into `development/2.5` 
* [w/2.6/bugfix/2634-impersonation-salt-api](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/bugfix/2634-impersonation-salt-api) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality

alexandre-allard commented on tests/post/steps/test_salt_api.py:
```diff
@@ -146,6 +146,12 @@ def run_state_on_targets(host, context, negated, module, targets):
 
+
+@then('authentication succeeds')
+def authentication_fails(host, context):
```
bad copy/paste ?
```suggestion
def authentication_succeeds(host, context):
```

gdemonet replied:
whoops

NicolasT commented on salt/_auth/kubernetes_rbac.py:
```diff
@@ -41,68 +43,101 @@ def wrapped(*args, **kwargs):
+    _patch_kubeconfig_for_basic(kubeconfig, username, password)
+
+    try:
+        access_review = _review_access(
```
Can't a user be valid, but not have access to the `SelfSubjectAccessReview` API?

gdemonet replied:
IIUC, any authenticated user will have the `system:authenticated` group. By default, this group is bound to the `system:basic-user` ClusterRole, which enables these permissions:

```yaml
rules:
- apiGroups:
  - authorization.k8s.io
  resources:
  - selfsubjectaccessreviews
  - selfsubjectrulesreviews
  verbs:
  - create
```

gdemonet replied:
Should I add this as a comment inline?

NicolasT replied:
Yes please.

gdemonet replied:
Will do!

gdemonet replied:
Done: 54aca8a1be857f6547f93f610f145af5e645d2d9

NicolasT commented on salt/_auth/kubernetes_rbac.py:
```diff
@@ -116,7 +151,34 @@ def _groups_basic(kubeconfig, username, token):
+        else:
+            return True
+    else:
+        log.error('Provided token failed to authenticate')
```
Can we log the provided `username` here?

gdemonet replied:
Done

NicolasT commented on salt/_auth/kubernetes_rbac.py:
```diff
@@ -148,23 +210,13 @@ def _load_kubeconfig(opts):
     return kubeconfig
 
 
+@_check_auth_args()
```
When calling with an incorrect combination of arguments, shouldn't this `raise` instead of return `False` as it does now?

gdemonet replied:
Done

NicolasT commented on salt/_auth/kubernetes_rbac.py:
```diff
@@ -184,22 +236,10 @@ def auth(username, password=None, token=None, **kwargs):
     return result
 
 
+@_check_auth_args(on_fail=lambda: [])
```
When calling with an incorrect combination of arguments, shouldn't this `raise` instead of return `[]` as it does now?

gdemonet replied:
Makes sense, and would make the code simpler :) I blindly followed the previous approach

gdemonet replied:
Done

NicolasT commented on tests/post/steps/test_salt_api.py:
```diff
@@ -54,30 +65,49 @@ def context():
     )
 
 
+@when(parsers.parse(
```
There's a fair bit of duplication between this procedure and the one above. Could be split out?

gdemonet replied:
Yes,  I'll cherry-pick the `_login_salt_api_sa` from 9d7e63d7d7f3e7601c0af02d2b1a61ae5ab73c82 (which I already customized in the 2.5 branch)

gdemonet replied:
Done (not a real cherry-pick, but backported the same logic/name)

NicolasT commented on ui/src/services/salt/api.js:
```diff
@@ -14,7 +14,7 @@ export function authenticate(user) {
   return saltApiClient.post('/login', {
     eauth: 'kubernetes_rbac',
     username: user.username,
-    password: user.token,
```
So where did `user.token` go? Should it still be set? Does it have a different meaning now?

gdemonet replied:
In 2.4, UI relies on basic auth. I didn't check the `user` state management in details, but both `password` and `token` are set. The API changed however, `password` isn't expecting a token anymore, but the password itself.

NicolasT replied:
Yes, but the `token` attribute on this `user` object now still exists. For no good reason?

gdemonet replied:
Indeed, I don't think it's used. OK will remove it, and get some more conflicts, yay :trollface:!

NicolasT replied:
Got `git rerere` enabled, right?

gdemonet replied:
Done: 94251f5

gdemonet replied:
> Got `git rerere` enabled, right?

Yes of course, but now I'm changing things that I didn't before :)

bert-e commented:
# History mismatch



Merge commit #82ed4189ea2aaa88dac47cb7c258f44a0f3a5989 on the integration branch
`w/2.5/bugfix/2634-impersonation-salt-api` is merging a branch which is neither the current
branch `bugfix/2634-impersonation-salt-api` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `bugfix/2634-impersonation-salt-api` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.5/bugfix/2634-impersonation-salt-api` with contents from `bugfix/2634-impersonation-salt-api`
and `development/2.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.5/bugfix/2634-impersonation-salt-api origin/development/2.5
 $ git merge origin/bugfix/2634-impersonation-salt-api
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.5/bugfix/2634-impersonation-salt-api
```

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/2634-impersonation-salt-api` into
`development/2.4`
* [w/2.5/bugfix/2634-impersonation-salt-api](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/bugfix/2634-impersonation-salt-api) will be merged into `development/2.5` 
* [w/2.6/bugfix/2634-impersonation-salt-api](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/bugfix/2634-impersonation-salt-api) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality

bert-e commented:
# History mismatch



Merge commit #82b164c4da79bc3721f69c0d5cee134cf2e0321d on the integration branch
`w/2.5/bugfix/2634-impersonation-salt-api` is merging a branch which is neither the current
branch `bugfix/2634-impersonation-salt-api` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `bugfix/2634-impersonation-salt-api` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

bert-e commented:
# History mismatch



Merge commit #4ade06db5151669efbb22606655df25c1256a7eb on the integration branch
`w/2.5/bugfix/2634-impersonation-salt-api` is merging a branch which is neither the current
branch `bugfix/2634-impersonation-salt-api` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `bugfix/2634-impersonation-salt-api` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality

gdemonet commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality










*The following options are set:* **approve**

gdemonet committed:
salt: Add comment about RBAC for access reviews

The fact that any authenticated user can use the SelfSubjectAccessReview
API was not self-evident. We add a comment to clarify this.

See: #2634

gdemonet committed:
ui: Remove storage of base64-encoded Basic token

We only need this token when interacting with K8s API, and it's already
stored in LocalStorage. We don't need to save it in the Redux store.

See: #2634

bert-e commented:
# Conflict




A conflict has been raised during the update of
integration branch `w/2.5/bugfix/2634-impersonation-salt-api` with contents from `bugfix/2634-impersonation-salt-api`
and `development/2.5`.

Please resolve the conflict on **the integration branch** (`w/2.5/bugfix/2634-impersonation-salt-api`).


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout w/2.5/bugfix/2634-impersonation-salt-api
 $ git pull  # or "git reset --hard origin/w/2.5/bugfix/2634-impersonation-salt-api"
 $ git merge origin/development/2.5
 $ # <intense conflict resolution>
 $ git commit
 $ git merge origin/bugfix/2634-impersonation-salt-api
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.5/bugfix/2634-impersonation-salt-api
```





*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality










*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`

* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`

* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue None.

Goodbye gdemonet.
